### PR TITLE
DAOS-9623 agent: Let client specify desired fabric iface

### PR DESCRIPTION
**Pull request by kjacque**
If the client has an OFI_INTERFACE/OFI_DOMAIN specified, the agent
should select the GetAttachInfo payload for the provider that matches
their choice.

- Include OFI_INTERFACE/OFI_DOMAIN in client GetAttachInfo request.
- If the requested interface/domain doesn't exist in the system,
  the primary provider is used.

Features: control

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

File: src/control/cmd/daos_agent/fabric.go [hunk 1]
```diff
@@ -273,6 +264,76 @@ func (n *NUMAFabric) setDefaultNUMANode() {
...
+		return result, nil
+	}
+
+	return nil, fmt.Errorf("fabric interface %q not found", name)
+}
+
+// FindDevice looks up a fabric device with a given name, domain, and provider.
+// NB: The domain and provider are optional. All other parameters are required. If there is more
+// than one match, all of them are returned.
+func (n *NUMAFabric) FindDevice(name, domain, provider string) ([]*FabricInterface, error) {
```
Comment (mjmac) [inline src/control/cmd/daos_agent/fabric.go:292]: Instead of having multiple string parameters, some of which are optional, I wonder if it might be cleaner design to switch over more of these methods to using the `fiParams` struct to provide the equivalent of named parameters. It's more verbose at the call sites, I guess, but seems less error-prone.

File: src/control/cmd/daos_agent/mgmt_rpc.go [hunk 2]
```diff
@@ -246,7 +293,19 @@ func (mod *mgmtModule) getFabricInterface(ctx context.Context, numaNode int, net
...
+}
+
+func (mod *mgmtModule) getCachedInterface(ctx context.Context, params *fiParams) (*FabricInterface, error) {
+	if params.interfaceName != "" {
+		fi, err := mod.fabricInfo.localNUMAFabric.FindDevice(params.interfaceName,
+			params.interfaceDomain, params.provider)
+		if err != nil {
+			return nil, err
+		}
+		return fi[0], nil
```
Comment (mjmac) [inline src/control/cmd/daos_agent/mgmt_rpc.go:306]: Will it always be correct to return the first one? Not fully following the logic here; maybe the thought is that any of them would be fine, so choosing the first is simplest?

File: src/mgmt/cli_mgmt.c [hunk 3]
```diff
@@ -243,9 +245,18 @@ get_attach_info(const char *name, bool all_ranks, struct dc_mgmt_sys_info *info,
 		D_GOTO(out, rc);
 	}
 
+	ofi_interface = getenv("OFI_INTERFACE");
+	if (ofi_interface)
+		D_INFO("Using client provided OFI_INTERFACE: %s\n", ofi_interface);
+	ofi_domain = getenv("OFI_DOMAIN");
+	if (ofi_domain)
+		D_INFO("Using client provided OFI_DOMAIN: %s\n", ofi_domain);
```
Comment (mjmac) [inline src/mgmt/cli_mgmt.c:253]: This change is to tell the agent about the client's desired network details, if they were set in the environment by the application user, right? What happens if the user sets some crazy stuff (e.g. copy/paste from some other env without understanding what they're for)?

I assume that there will be some logging on the agent side, but wondering how the error(s) will manifest to the application user.

File: src/control/cmd/daos_agent/mgmt_rpc.go (hunk 2 above)
Comment (kjacque) [reply src/control/cmd/daos_agent/mgmt_rpc.go:306]: That's the thought, yeah. If the interface, domain, and provider match, it's sufficient. In practice we'll probably only get one for the combination of all three params, but I wanted this function to be flexible in the other area where we use it (line 198).

File: src/mgmt/cli_mgmt.c (hunk 3 above)
Comment (kjacque) [reply src/mgmt/cli_mgmt.c:253]: The primary goal of these changes is to choose the correct provider URIs/hint based on the client's env variables. I.e. pick a secondary provider if that's all the requested iface/domain enables. Initially I was planning to do more to validate the iface/domain, but after a couple conversations last week, I came to the conclusion that we shouldn't validate, and instead let them blindly override like before.

Basically I'm thinking of cases where we're adding new providers that aren't detected by the agent yet, like the recent development on UCX support. Developers are the only ones who should really be using these env variables.

File: src/control/cmd/daos_agent/fabric.go (hunk 1 above)
Comment (kjacque) [reply src/control/cmd/daos_agent/fabric.go:292]: Agreed, I think that would be cleaner.

File: src/mgmt/cli_mgmt.c (hunk 3 above)
Comment (mjmac) [reply src/mgmt/cli_mgmt.c:253]: Hmm. I'm a little skeptical about the assertion that only developers will set the variables. Users are known for doing strange and unexpected things (whether they mean to or not). I was thinking that the agent config could be used to override the autoconfig behavior, but maybe I'm not remembering some details.

Comment (kjacque) [reply src/mgmt/cli_mgmt.c:253]: Added some logging like we discussed, and I ended up moving things around a little. The errors printed to the agent log should at least provide a hint to folks doing something they're not supposed to.

File: src/control/cmd/daos_agent/mgmt_rpc.go [hunk 4]
```diff
@@ -143,30 +144,45 @@ func (mod *mgmtModule) getNUMANode(ctx context.Context, pid int32) (uint, error)
...
 	}
 
-	fabricIF, err := mod.getFabricInterface(ctx, numaNode, hardware.NetDevClass(resp.ClientNetHint.NetDevClass),
-		resp.ClientNetHint.Provider)
-	if err != nil {
-		mod.log.Errorf("failed to fetch fabric interface of type %s: %s",
-			hardware.NetDevClass(resp.ClientNetHint.NetDevClass), err.Error())
-		return nil, err
+	// Requested fabric interface/domain behave as a simple override. If we weren't able to
+	// validate them, we return them to the user with the understanding that perhaps the user
```
Comment (tanabarr) [inline src/control/cmd/daos_agent/mgmt_rpc.go:162]: I like this proposition :-)

File: src/control/cmd/daos_agent/mgmt_rpc.go [hunk 5]
```diff
@@ -178,35 +194,80 @@ func (mod *mgmtModule) getAttachInfoResp(ctx context.Context, numaNode int, sys
...
+	if domain == "" {
+		domain = iface
+	}
+
+	fis, err := mod.fabricInfo.localNUMAFabric.FindDevice(&FabricIfaceParams{
+		Interface: iface,
+		Domain:    domain,
+	})
+	if err != nil {
+		mod.log.Errorf("client-requested fabric interface/domain not detected: %s", err.Error())
```
Comment (tanabarr) [inline src/control/cmd/daos_agent/mgmt_rpc.go:211]: Why make this 2 separate error entries in the same log?

File: src/control/cmd/daos_agent/mgmt_rpc.go [hunk 6]
```diff
@@ -178,35 +194,80 @@ func (mod *mgmtModule) getAttachInfoResp(ctx context.Context, numaNode int, sys
...
+		providers.AddUnique(fi.Providers()...)
+	}
+	return providers
+}
+
+func (mod *mgmtModule) selectAttachInfo(srvResp *mgmtpb.GetAttachInfoResp, reqProviders common.StringSet) (*mgmtpb.GetAttachInfoResp, error) {
+	providers := reqProviders
+	if mod.provider != "" {
+		if len(reqProviders) > 0 && !reqProviders.Has(mod.provider) {
+			mod.log.Errorf("configured provider %q not included in requested interface's detected providers: %s", reqProviders)
```
Comment (tanabarr) [inline src/control/cmd/daos_agent/mgmt_rpc.go:227]: same as above, I'm guessing you just want the "communications on this interface may fail" to be an independent entry?

**Review by tanabarr — APPROVED** [review]

File: src/control/cmd/daos_agent/fabric.go [hunk 7]
```diff
@@ -273,6 +277,80 @@ func (n *NUMAFabric) setDefaultNUMANode() {
...
+			return nil, errors.Errorf("fabric interface %q doesn't support provider %q",
+				params.Interface, params.Provider)
+		}
+	}
+
+	return fiList, nil
+}
+
+func filterDomain(domain string, fiList []*FabricInterface) []*FabricInterface {
+	result := make([]*FabricInterface, 0, len(fiList))
```
Comment (knard-intel) [inline src/control/cmd/daos_agent/fabric.go:335]: Nitpicking, but we are probably going to do some overprovisioning as all the elements of the `filist` will probably not be added to result.

File: src/control/cmd/daos_agent/fabric.go [hunk 8]
```diff
@@ -273,6 +277,80 @@ func (n *NUMAFabric) setDefaultNUMANode() {
...
+	for _, fi := range fiList {
+		if fi.Domain == domain || (fi.Name == domain && fi.Domain == "") {
+			result = append(result, fi)
+		}
+	}
+	return result
+}
+
+func filterProvider(provider string, fiList []*FabricInterface) []*FabricInterface {
+	result := make([]*FabricInterface, 0, len(fiList))
```
Comment (knard-intel) [inline src/control/cmd/daos_agent/fabric.go:345]: Same comment for the overprovisioning of the slice.

**Review by knard-intel — APPROVED** [review]
Seems good for what I understand.

File: src/control/cmd/daos_agent/mgmt_rpc.go (hunk 6 above)
Comment (mjmac) [reply src/control/cmd/daos_agent/mgmt_rpc.go:227]: Yes, the idea was that we want to make this somewhat noisy so that if there are communications failures due to invalid parameters supplied via env variables it's not a mystery. We don't want the env variables to be used as a primary configuration method for users, hence the relaxed approach to validation ("okay, if you think this will work we'll try it, but as far as we know it's busted").

Comment (kjacque) [reply src/control/cmd/daos_agent/mgmt_rpc.go:227]: Yeah, my goal was to make it a bit loud. I wanted to clarify what the consequences of the error are in case someone comes looking.

File: src/control/cmd/daos_agent/fabric.go (hunk 7 above)
Comment (kjacque) [reply src/control/cmd/daos_agent/fabric.go:335]: Fair point, although we're not overprovisioning by much, as each network interface has 2-3 domains at most.

**Review by mjmac — APPROVED** [review]
